### PR TITLE
[expo-linear-gradient] Require at least two colors in typescript

### PR DESCRIPTION
**Pull request by navignaw**
# Why

If you pass less than 2 colors in LinearGradient, the generated React Native Android code crashes.

```
java.lang.IllegalArgumentException: needs >= 2 number of colors
	at android.graphics.Shader.convertColors(Shader.java:215)
	at android.graphics.LinearGradient.<init>(LinearGradient.java:71)
	at expo.modules.lineargradient.LinearGradientView.drawGradient(LinearGradientView.java:83)
	at expo.modules.lineargradient.LinearGradientView.onSizeChanged(LinearGradientView.java:76)
	at android.view.View.sizeChange(View.java:24612)
	at android.view.View.setFrame(View.java:24564)
	at android.view.View.layout(View.java:24415)
	at com.facebook.react.uimanager.NativeViewHierarchyManager.updateLayout(NativeViewHierarchyManager.java:254)
	at com.facebook.react.uimanager.NativeViewHierarchyManager.updateLayout(NativeViewHierarchyManager.java:222)
	at com.swmansion.reanimated.layoutReanimation.ReanimatedNativeHierarchyManager.updateLayout(ReanimatedNativeHierarchyManager.java:256)
	at com.facebook.react.uimanager.UIViewOperationQueue$UpdateLayoutOperation.execute(UIViewOperationQueue.java:169)
	at com.facebook.react.uimanager.UIViewOperationQueue$1.run(UIViewOperationQueue.java:926)
	at com.facebook.react.uimanager.UIViewOperationQueue.flushPendingBatches(UIViewOperationQueue.java:1037)
	at com.facebook.react.uimanager.UIViewOperationQueue.-$$Nest$mflushPendingBatches(Unknown Source:0)
	at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.doFrameGuarded(UIViewOperationQueue.java:1097)
	at com.facebook.react.uimanager.GuardedFrameCallback.doFrame(GuardedFrameCallback.java:29)
	at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:175)
	at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame(ChoreographerCompat.java:85)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1337)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1348)
	at android.view.Choreographer.doCallbacks(Choreographer.java:952)
	at android.view.Choreographer.doFrame(Choreographer.java:878)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1322)
	at android.os.Handler.handleCallback(Handler.java:958)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8177)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
```

# How

In order to prevent this, we can statically typecheck that at least two colors are passed into the array. The solution for requiring at least two values was taken from https://stackoverflow.com/a/57117431

# Test Plan

Attempt to import this component and pass fewer than two colors to it. Verify that TS does not pass.

```tsx
import { LinearGradient } from 'expo-linear-gradient';

// successfully typechecks
const colors = ['cyan', '#ff00ff', 'rgba(0,0,0,0)', 'rgba(0,255,255,0.5)'] as const;
return (
  <LinearGradient
    colors={colors}
    style={styles.background}
  />
)

// successfully typechecks
return (
  <LinearGradient
    colors={['rgba(0,0,0,0.8)', 'transparent']}
    style={styles.background}
  />
)

// fails to typecheck
return (
  <LinearGradient
    colors={['#ffffff']}
    style={styles.background}
  />
)
```

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by navignaw** [comment]
Opening this PR in draft mode while I figure out how to actually test this. Any advice would be appreciated 😅

**Review by vonovak — CHANGES_REQUESTED** [review]
Hello and thank you for your PR!
Could you change the PR so that 

(1) the target branch is `main`
(2) you run `et gdad -p expo-linear-gradient` in the repo, and commit the result of that as well?

Thank you!

**Comment by navignaw** [comment]
> (1) the target branch is `main` (2) you run `et gdad -p expo-linear-gradient` in the repo, and commit the result of that as well?

Done!

**Review by vonovak — COMMENTED** [review]
Hello and thanks for making those changes. There's one issue on our side now, which is that the docs are not properly generated. As a result, there's no entry for `LinearGradientColors` in the generated docs. 

As for the code, one things that might be cumbersome is that

this:

```tsx
  const colors = ['cyan', '#ff00ff', 'rgba(0,0,0,0)', 'rgba(0,255,255,0.5)'];
  render(<LinearGradient colors={colors} testID="gradient" />);
```

needs to become typed with `LinearGradientColors` or `as const`

```tsx
  const colors: LinearGradientColors = ['cyan', '#ff00ff', 'rgba(0,0,0,0)', 'rgba(0,255,255,0.5)'];
  render(<LinearGradient colors={colors} testID="gradient" />);
```

Personally, I'm fine with that but will ask for another opinion. At any rate, we first need to fix the docs output. I'll get back to this when that's done.

Thank you for your PR!

File: packages/expo-linear-gradient/src/LinearGradient.tsx [hunk 1]
```diff
@@ -33,7 +38,7 @@ export type LinearGradientProps = ViewProps & {
    * A readonly array of colors that represent stops in the gradient. At least two colors are required
    * (for a single-color background, use the `style.backgroundColor` prop on a `View` component).
    */
-  colors: readonly string[];
+  colors: LinearGradientColors;
```
Comment (tsapeta) [inline packages/expo-linear-gradient/src/LinearGradient.tsx:41]: I believe this works too and maybe the docs output is a bit better, @vonovak?
```suggestion
  colors: readonly [string, string, ...string[]];
```

**Comment by navignaw** [comment]
> As for the code, one things that might be cumbersome is that
> 
> this:
> 
> ```tsx
>   const colors = ['cyan', '#ff00ff', 'rgba(0,0,0,0)', 'rgba(0,255,255,0.5)'];
>   render(<LinearGradient colors={colors} testID="gradient" />);
> ```
> 
> needs to become typed with `LinearGradientColors` or `as const`
> 
> ```tsx
>   const colors: LinearGradientColors = ['cyan', '#ff00ff', 'rgba(0,0,0,0)', 'rgba(0,255,255,0.5)'];
>   render(<LinearGradient colors={colors} testID="gradient" />);
> ```

Yeah, I was definitely on the fence with this. It's not an issue if you define the colors inline because TSX is smart enough to infer if there are enough colors, but it will probably be a pain point for users who define an array and then pass it in, even if that array already has >= 2 values.

I still think that the value of catching this statically is far superior a debugging experience than catching it live in production though. It's just a question of how to introduce this "breaking change" to existing users.

The reason I defined `LinearGradientColors` in the first place (rather than inlining it) was so that it would be easy to import it and use it to annotate existing arrays. But I'm open to other suggestions!

File: packages/expo-linear-gradient/src/LinearGradient.tsx (hunk 1 above)
Comment (vonovak) [reply packages/expo-linear-gradient/src/LinearGradient.tsx:41]: @tsapeta the first problem seems to be readonly - even [the current docs](https://docs.expo.dev/versions/latest/sdk/linear-gradient/#colors) don't render it properly.

this: `colors: [string, string, ...string[]];` renders as below, so it's also a bit odd

Screenshot 2024-08-08 at 9 43 59

**Review by vonovak — APPROVED** [review]
Even if we don't end up merging the change to typings, we should pick the changes that fix rendering of `readonly` props (and tuples):


Screenshot 2024-08-08 at 12 48 15

**Comment by navignaw** [comment]
Thanks for fixing the example types! I do feel that if we don't expose the `LinearGradientColors` we might want to include a blurb in the docs explaining to users how to address any type issues that come up. How do you feel about something like this?

> Due to a limitation in TS, any dynamically typed variables need to be explicitly typed for TS to know it has more than 2 values. For example:
> ```ts
> const colors: [string, string, ...string[]] = ['red', 'blue', 'green'] as const
> ```

**Comment by vonovak** [comment]
> Thanks for fixing the example types! I do feel that if we don't expose the `LinearGradientColors` we might want to include a blurb in the docs explaining to users how to address any type issues that come up. How do you feel about something like this?
> 
> > Due to a limitation in TS, any dynamically typed variables need to be explicitly typed for TS to know it has more than 2 values. For example:
> > ```ts
> > const colors: [string, string, ...string[]] = ['red', 'blue', 'green'] as const
> > ```

Yeah, thought about it too. I'm for adding some note, but not too long. Maybe 

For TS to know the provided array has 2 or more values, it needs to be provided "inline" or typed `as const`. For example:
 
```ts
 const colors = ['red', 'blue', 'green'] as const
 ```

Either `const colors: [string, string, ...string[]] = ['red', 'blue', 'green']` or `const colors: [string, string, ...string[]] = ['red', 'blue', 'green']` should be the used syntax. 

`const colors: [string, string, ...string[]] = ['red', 'blue', 'green'] as const` actually isn't valid (it'd need to be `const colors: readonly [string, string, ...string[]] = ['red', 'blue', 'green'] as const`). 

To avoid too much type gymnastics, I'd suggest using `as const`.

**Comment by navignaw** [comment]
Good point! I like the `as const` the most since it gets around having to define the somewhat ugly `[string, string, ...string[]]` type.

**Review by tsapeta — APPROVED** [review]

**Comment by navignaw** [comment]
Any idea why the docs pr action is failing on `Unable to locate credentials. You can configure credentials by running "aws configure".`?

**Comment by navignaw** [comment]
> Yeah, thought about it too. I'm for adding some note, but not too long. Maybe
> 
> For TS to know the provided array has 2 or more values, it needs to be provided "inline" or typed `as const`. For example:
> 
> ```ts
>  const colors = ['red', 'blue', 'green'] as const
> ```

Added this to the docs!

File: packages/expo-linear-gradient/src/LinearGradient.tsx [hunk 2]
```diff
@@ -78,7 +78,7 @@ export type LinearGradientProps = ViewProps & {
 export class LinearGradient extends React.Component<LinearGradientProps> {
   render() {
     const { colors, locations, start, end, dither, ...props } = this.props;
-    let resolvedLocations = locations;
+    let resolvedLocations = locations as number[] | undefined;
```
Comment (Simek) [inline packages/expo-linear-gradient/src/LinearGradient.tsx:81]: Nit, prefer typing over casting. In this case we miss `undefined` in there.

```suggestion
    let resolvedLocations: number[] | null | undefined = locations;
```

or

```suggestion
    let resolvedLocations: LinearGradientProps['locations']  = locations;
```

**Review by Simek — APPROVED** [review]
Thanks for adding missing renderers for `readOnly` and rest spread in parameters, good work! 👍 

Let's make sure before shipping that altering `type === 'typeOperator'` branch logic di not affect existing packages data negatively. It looks like this code branch have been used before to render `keyOf` values.

Also, the docs changes contains samll lint issue, let's address that too.

File: packages/expo-linear-gradient/src/LinearGradient.tsx (hunk 2 above)
Comment (navignaw) [reply packages/expo-linear-gradient/src/LinearGradient.tsx]: Thanks, had to apply an extra `readonly` for it to typecheck.

**Comment by navignaw** [comment]
Fixed the lint error and the types!

> Let's make sure before shipping that altering type === 'typeOperator' branch logic di not affect existing packages data negatively. It looks like this code branch have been used before to render keyOf values.

It looks like that was moved into #30954. I just merged upstream and this PR should only touch files related to linear gradient now.

File: docs/pages/versions/unversioned/sdk/linear-gradient.mdx [hunk 3]
```diff
@@ -79,4 +79,10 @@ const styles = StyleSheet.create({
 import { LinearGradient } from 'expo-linear-gradient';
 ```
 
+For TS to know the provided array has 2 or more values, it needs to be provided "inline" or typed `as const`. For example:
```
Comment (vonovak) [inline docs/pages/versions/unversioned/sdk/linear-gradient.mdx:82]: can you move this directly into the docs for the `color` prop? It'll be better positioned there I think. They are generated from packages/expo-linear-gradient/src/LinearGradient.tsx

File: packages/expo-linear-gradient/CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 ### 🛠 Breaking changes
 
+- The colors prop now requires at least two values. Passing an empty array or an array with one value will fail to typecheck. ([#27820](https://github.com/expo/expo/pull/27820) by [@navignaw](https://github.com/navignaw))
```
Comment (vonovak) [inline packages/expo-linear-gradient/CHANGELOG.md:7]:
```suggestion
- stricter TS type of the `colors` prop. It now requires providing an array of least two values. Passing an empty array or an array with one value will fail to typecheck. ([#27820](https://github.com/expo/expo/pull/27820) by [@navignaw](https://github.com/navignaw))
```

the `colors` prop always required >= 2 values so this is more accurate

**Review by vonovak — CHANGES_REQUESTED** [review]

**Comment by aleqsio** [comment]
I know this is kinda late in this PR, but is the expected behaviour to require 2 colors in TS? 
I'd argue that a `<LinearGradient colors={["red"]} />` should just be a red background, and not crash 😅 .

**Comment by vonovak** [comment]
> I know this is kinda late in this PR, but is the expected behaviour to require 2 colors in TS? I'd argue that a `<LinearGradient colors={["red"]} />` should just be a red background, and not crash 😅 .
> 
> The crash is coming from https://android.googlesource.com/platform/frameworks/base/+/master/graphics/java/android/graphics/Shader.java, which already seems like a rather weird assumption, but fixing it would require us to fallback to not rendering a gradient view.
> 

Google chose it this way which may be odd, but I don't think that we should fall back to rendering something else than a GradientView if user wants a GradientView, it'd feel like too much magic to me. It's true that JS users don't benefit this, but TS users do and that's already an improvement.

RN core is getting support for linear gradients [1](https://github.com/facebook/react-native/pull/45434), [2](https://github.com/facebook/react-native/pull/45433) so at some point I don't think we'll need this, but if there's a chance for improving what we have, why not take it. 🙂

**Review by vonovak — APPROVED** [review]

**Comment by vonovak** [comment]
@navignaw thanks for the PR and for your patience! 💯

**Comment by Noitham** [comment]
@vonovak - I think these new types gives conflicts when having types defined for a color scheme and gradients are defined as string[].

My color schemes are based on string[], and although I can see I'm always passing 2 values to the array string, I'm seeing this error when using the LinearGradient component:

```
No overload matches this call.
  Overload 1 of 2, '(props: LinearGradientProps): LinearGradient', gave the following error.
    Type 'string[]' is not assignable to type 'readonly [string, string, ...string[]]'.
  Overload 2 of 2, '(props: LinearGradientProps, context: any): LinearGradient', gave the following error.
    Type 'string[]' is not assignable to type 'readonly [string, string, ...string[]]'
```

**Comment by navignaw** [comment]
> @vonovak - I think these new types gives conflicts when having types defined for a color scheme and gradients are defined as string[].
> 
> My color schemes are based on string[], and although I can see I'm always passing 2 values to the array string, I'm seeing this error when using the LinearGradient component:
> 
> ```
> No overload matches this call.
>   Overload 1 of 2, '(props: LinearGradientProps): LinearGradient', gave the following error.
>     Type 'string[]' is not assignable to type 'readonly [string, string, ...string[]]'.
>   Overload 2 of 2, '(props: LinearGradientProps, context: any): LinearGradient', gave the following error.
>     Type 'string[]' is not assignable to type 'readonly [string, string, ...string[]]'
> ```
> 
> CC: @navignaw

@Noitham take a look at this comment: https://github.com/expo/expo/pull/27820#issuecomment-2276156022 and the [docs](https://docs.expo.dev/versions/latest/sdk/linear-gradient/#component)

> For TypeScript to know the provided array has 2 or more values, it should be provided "inline" or typed `as const`.

Unfortunately this is a limitation of TypeScript as it has no idea whether the array you're passing in has at least two strings. If you define your variable `const myColors = [..., ...] as const` then it will force it to be readonly and check that it has enough strings.

If you are still having trouble feel free to provide the code where you define the colors and I'm happy to help take a look

**Comment by navignaw** [comment]
Hm not sure if there is a bug in the documentation code or if it got reverted, but a prior commit gave an example for the colors prop which I am no longer seeing in https://docs.expo.dev/versions/latest/sdk/linear-gradient/#component

https://github.com/expo/expo/commit/d613c9c46262feda23068b4e2c9acfde23bb28f4

**Pull request merged**